torch/utils/_pytree.py

```diff
@@ -26,194 +28,124 @@
 To improve the performance we can move parts of the implementation to C++.
 """
 
+PROTOCOL_VERSION = 1
+
+Context = Any
+PyTree = Any
+FlattenFunc = Callable[[PyTree], Tuple[List, Context]]
+UnflattenFunc = Callable[[List, Context], PyTree]
+DumpableContext = Any  # Any json dumpable text
+GetStateFn = Callable[[Context], DumpableContext]
+SetStateFn = Callable[[DumpableContext], Context]
+
 # A NodeDef holds two callables:
 # - flatten_fn should take the collection and return a flat list of values.
 #   It can also return some context that is used in reconstructing the
 #   collection.
 # - unflatten_fn should take a flat list of values and some context
 #   (returned by flatten_fn). It returns the collection by reconstructing
 #   it from the list and the context.
-# - to_str_fn takes a TreeSpec with the specific type and a list of its children
-#   TreeSpecs already converted to strings, and returns a string representation
-#   of this TreeSpec
-# - maybe_from_str_fn takes in a string and if this string represents a TreeSpec
-#   of this type, returns the type, the context, and a string representation of
-#   its children specs. Otherwise it returns None.
-Context = Any
-PyTree = Any
-FlattenFunc = Callable[[PyTree], Tuple[List, Context]]
-UnflattenFunc = Callable[[List, Context], PyTree]
-ToStrFunc = Callable[["TreeSpec", List[str]], str]
-MaybeFromStrFunc = Callable[[str], Optional[Tuple[Any, Context, str]]]
-
 class NodeDef(NamedTuple):
     type: Type[Any]
     flatten_fn: FlattenFunc
     unflatten_fn: UnflattenFunc
-    to_str_fn: ToStrFunc
-    maybe_from_str_fn: MaybeFromStrFunc
 
-SUPPORTED_NODES: Dict[Type[Any], NodeDef] = {}
+SUPPORTED_NODES: Dict[Union[Type[Any], str], NodeDef] = {}
+
+# SerializeNodeDef holds two callables:
+# - serialize_fn takes a TreeSpec, and returns a serialized string format of the
+#   context, and the version number
+# - deserialize_fn takes in a string representation of the context, and the
+#   version, and returns the deserialized context
+class SerializeNodeDef(NamedTuple):
+    typ: Type[Any]
+    serialized_type: str
+    getstate: Optional[GetStateFn]
+    setstate: Optional[SetStateFn]
+
+SUPPORTED_SERIALIZED_TYPES: Dict[Union[Type[Any], str], SerializeNodeDef] = {}
+
+def _register_treespec_serializer(
+    typ: Type[Any],
+    *,
+    getstate: Optional[GetStateFn] = None,
+    setstate: Optional[SetStateFn] = None,
+) -> None:
+    serialized_type = f"{typ.__module__}.{typ.__name__}"
+    serialize_node_def = SerializeNodeDef(
+        typ, serialized_type, getstate, setstate
+    )
+    SUPPORTED_SERIALIZED_TYPES[typ] = serialize_node_def
+    SUPPORTED_SERIALIZED_TYPES[serialized_type] = serialize_node_def
```

Comment:
Why indexed by both? We should just do one

Reply:
I couldn't think of how to get the type back after serializing it...I could do some python importing like, but I think it doesn't work for testing purposes:

```
for module_name in type_str.split("."):
    if current_module is None:
        current_module = importlib.import_module(module_name)
    else:
        current_module = getattr(current_module, module_name)
```

Reply:
Yeah, we shouldn't be in the business of doing python imports on treespec_load.

Here's an idea:
- there should be a Dict somewhere that maps fully qualified name to the type
- on treespec_loads, if the FQN isn't in this Dict, then that means the user has not imported the file that adds pytree support for the type yet and we should raise an error.

Ugh there's also the possiblity of desync issues, like if someone has a `_register_pytree_impl` but then does a separate `_register_treespec_serializer` in another file. I know I suggested splitting the functions, but maybe we actually need to put them back together to avoid desync issues (and make getstate/setstate kwarg-only so this is extensible in the future)

Reply:
> there should be a Dict somewhere that maps fully qualified name to the type

Ok this makes sense!

> on treespec_loads, if the FQN isn't in this Dict, then that means the user has not imported the file that adds pytree support for the type yet and we should raise an error.

This is being done rn

> we actually need to put them back together to avoid desync issues

Yeah..I thought about it and I was just hoping that someone would call `_register_pytree_impl` before `_register_treespec_serializer` 😅. I can fix it